File: dashboard/server/master_data.js
/**
 * Preserve the former Node master-data derivation as a parity fixture.
 *
 * The Budget Manager's entity sections -- Ad Providers, Products, Campaigns,
 * Ad Groups, Touchpoints, Product Economics -- describe the account the
 * reports were pulled from. Those reports are tracked in the repository, so
 * the catalogue they imply is available in every deployment and does not
 * depend on any optional sidecar being configured: the entity aggregate names
 * every Campaign, Ad Group, ASIN, and SKU, and the daily platform report names
 * every touchpoint with its billing type and observed cost.
 *
 * Derivation rather than a second tracked file is deliberate. A catalogue
 * committed beside the reports would be free to disagree with them; one read
 * out of the reports cannot. Where a report does not carry a field -- a
 * Product's category, a unit price, a Campaign's baseline budget -- the field
 * stays null rather than being invented, and the interface renders it as
 * missing.
 *
 * Runtime requests use `backend/repository/master_data.py`. This implementation
 * remains the cross-language reference imported by `server/data_source.js`.
 *
 * Parity-test flow:
 *     committed reports -> here -> server/data_source.js -> dashboard/tests
 */

/** Sum a numeric column across rows, treating blanks as zero. */
function total(rows, field) {
  return rows.reduce((sum, row) => sum + (Number(row[field]) || 0), 0);
}

/** Distinct non-empty values of `field`, in first-seen order. */
function distinct(rows, field) {
  const seen = [];
  for (const row of rows) {
    const value = row[field];
    if (value !== undefined && value !== null && value !== "" && !seen.includes(value)) {
      seen.push(value);
    }
  }
  return seen;
}

/**
 * The five-segment touchpoint key's parts.
 *
 * `UNSPECIFIED` is the pipeline's marker for a segment the platform does not
 * report, and is mapped back to null so the interface shows it as absent
 * rather than as a literal value the account actually uses.
 */
function splitKey(key) {
  const [adProduct, format, placement, creative, interaction] = String(key).split(":");
  const clean = (value) => (value && value !== "UNSPECIFIED" ? value : null);
  return {
    ad_product: adProduct ?? null,
    format: clean(format),
    placement: clean(placement),
    creative: clean(creative),
    interaction_type: interaction ?? null,
  };
}

/**
 * The provider a five-segment key belongs to.
 *
 * The four Amazon ad products are the vocabulary `mta_common` defines for
 * `Provider.AMAZON_ADS`; anything else is reported under `GENERIC` rather
 * than being dropped, so an added provider surfaces instead of vanishing.
 */
const AMAZON_AD_PRODUCTS = new Set([
  "SPONSORED_PRODUCTS",
  "SPONSORED_BRANDS",
  "SPONSORED_DISPLAY",
  "AMAZON_DSP",
]);

function providerOf(adProduct) {
  return AMAZON_AD_PRODUCTS.has(adProduct) ? "AMAZON_ADS" : "GENERIC";
}

/** `AVAILABLE` when any row reports the segment, `NOT_PROVIDED` when none do. */
function availability(rows, field) {
  return rows.some((row) => row[field]) ? "AVAILABLE" : "NOT_PROVIDED";
}

/**
 * Build the catalogue.
 *
 * `adsRows` is the daily platform report and `bridgeRows` the touchpoint-to-
 * entity aggregate; both are already coerced by the loaders that read them.
 * Returns the same shape the research sidecar and the database produce, so
 * the views cannot tell which source filled it.
 */
export function deriveMasterData(adsRows, bridgeRows, strategyRequest = {}) {
  const touchpoints = deriveTouchpoints(adsRows);
  const campaigns = deriveCampaigns(bridgeRows, touchpoints, strategyRequest);
  // The account reports one currency; it is read from the data rather than
  // assumed, so a non-USD account is not silently relabelled.
  const currency =
    distinct(adsRows, "currency")[0] ??
    strategyRequest.campaign_group?.currency ??
    null;
  return {
    providers: deriveProviders(touchpoints),
    products: deriveProducts(bridgeRows),
    campaigns,
    adGroups: deriveAdGroups(bridgeRows),
    touchpoints,
    productEconomics: deriveProductEconomics(bridgeRows, currency),
    campaignProductLinks: deriveCampaignProductLinks(bridgeRows),
  };
}

/**
 * One record per distinct five-segment key, with its observed economics.
 *
 * Cost per click and cost per thousand impressions are computed from the
 * report's own totals rather than assumed, and each is reported only for the
 * billing type that produced it: a CPM touchpoint has no meaningful cost per
 * click even when a stray click was recorded against it.
 */
function deriveTouchpoints(adsRows) {
  const byKey = new Map();
  for (const row of adsRows) {
    const key = row.touchpoint ?? row.normalizedTouchpoint;
    if (!key) continue;
    if (!byKey.has(key)) byKey.set(key, []);
    byKey.get(key).push(row);
  }

  const byIdentifier = new Map();
  for (const [key, rows] of byKey) {
    const parts = splitKey(key);
    // The identifier drops the interaction segment: IMPRESSION and CLICK on
    // one placement are two interactions of a single touchpoint, which is the
    // grouping `supported_interactions` exists to express.
    const identifier = [
      parts.ad_product,
      parts.format ?? "UNSPECIFIED",
      parts.placement ?? "UNSPECIFIED",
      parts.creative ?? "UNSPECIFIED",
    ].join(":");

    const existing = byIdentifier.get(identifier);
    const record = existing ?? {
      identifier,
      provider: providerOf(parts.ad_product),
      ad_product: parts.ad_product,
      format: parts.format,
      placement: parts.placement,
      placement_availability: parts.placement ? "AVAILABLE" : "NOT_PROVIDED",
      creative: parts.creative,
      creative_availability: parts.creative ? "AVAILABLE" : "NOT_PROVIDED",
      interaction_type_availability: "AVAILABLE",
      supported_interactions: [],
      impression_enabled: false,
      click_enabled: false,
      billing_type: null,
      cost_per_click: null,
      cost_per_thousand_impressions: null,
      base_impressions: null,
      click_through_rate: null,
      platform_conversion_rate: null,
      conversion_log_odds_effect: null,
      compatibility_keys: [],
      active: true,
    };

    if (parts.interaction_type && !record.supported_interactions.includes(parts.interaction_type)) {
      record.supported_interactions.push(parts.interaction_type);
    }
    record.impression_enabled = record.supported_interactions.includes("IMPRESSION");
    record.click_enabled = record.supported_interactions.includes("CLICK");
    record.compatibility_keys.push(key);

    const impressions = total(rows, "impressions");
    const clicks = total(rows, "clicks");
    const cost = total(rows, "cost");

    // One touchpoint is commonly billed both ways -- CPC on its clicks and CPM
    // on its impressions -- so cost is accumulated against the billing type
    // that produced it rather than against the touchpoint as a whole. Dividing
    // a mixed total by either denominator would report a rate the platform
    // never charged.
    const billed = (record._billed ??= {});
    for (const row of rows) {
      const type = row.cost_type;
      if (!type) continue;
      const bucket = (billed[type] ??= { cost: 0, clicks: 0, impressions: 0 });
      bucket.cost += Number(row.cost) || 0;
      bucket.clicks += Number(row.clicks) || 0;
      bucket.impressions += Number(row.impressions) || 0;
    }

    const totals = (record._totals ??= { impressions: 0, clicks: 0, cost: 0 });
    totals.impressions += impressions;
    totals.clicks += clicks;
    totals.cost += cost;

    byIdentifier.set(identifier, record);
  }

  return [...byIdentifier.values()].map(({ _totals, _billed, ...record }) => {
    const types = Object.keys(_billed ?? {}).sort();
    const cpc = _billed?.CPC;
    const cpm = _billed?.CPM;
    return {
      ...record,
      // Both types when the platform bills both, so the row states what it is
      // rather than naming one and hiding the other.
      billing_type: types.length > 0 ? types.join(" + ") : null,
      // A rate only where cost was actually charged against that denominator.
      // An impression row that delivered volume at no recorded cost yields no
      // rate rather than a rate of zero, which would read as "free".
      cost_per_click: cpc?.cost > 0 && cpc.clicks > 0 ? round(cpc.cost / cpc.clicks, 4) : null,
      cost_per_thousand_impressions:
        cpm?.cost > 0 && cpm.impressions > 0
          ? round((cpm.cost / cpm.impressions) * 1000, 4)
          : null,
      observed_impressions: _totals?.impressions ?? 0,
      observed_clicks: _totals?.clicks ?? 0,
      observed_cost: round(_totals?.cost ?? 0, 2),
      // Deliberately not derived. The report records impressions and clicks in
      // separate per-interaction rows that share no denominator, so dividing
      // one by the other does not produce this touchpoint's click-through
      // rate; the observed counts above are what the platform actually states.
      click_through_rate: null,
      base_impressions: _totals?.impressions ?? null,
    };
  });
}

/** Round to a fixed precision without carrying float noise into the interface. */
function round(value, digits) {
  const factor = 10 ** digits;
  return Math.round(value * factor) / factor;
}

/**
 * One record per provider actually present in the reports.
 *
 * Each provider's supported ad products and per-segment availability are read
 * from its own touchpoints, so a provider that never reports a placement is
 * described as not providing one rather than as having none.
 */
function deriveProviders(touchpoints) {
  const byProvider = new Map();
  for (const touchpoint of touchpoints) {
    if (!byProvider.has(touchpoint.provider)) byProvider.set(touchpoint.provider, []);
    byProvider.get(touchpoint.provider).push(touchpoint);
  }
  return [...byProvider.entries()].map(([provider, rows]) => ({
    provider,
    supported_ad_products: distinct(rows, "ad_product"),
    format_availability: availability(rows, "format"),
    placement_availability: availability(rows, "placement"),
    creative_availability: availability(rows, "creative"),
    interaction_type_availability: "AVAILABLE",
    active: true,
  }));
}

/**
 * One record per advertised SKU.
 *
 * The reports identify a product by ASIN and SKU and carry nothing else about
 * it: no name, category, brand, or inventory. Those fields are present and
 * null, which is what lets the Form editor show the whole shape of a Product
 * while being honest that the platform report does not describe it.
 */
function deriveProducts(bridgeRows) {
  const bySku = new Map();
  for (const row of bridgeRows) {
    const sku = row.sku_id;
    if (!sku || bySku.has(sku)) continue;
    bySku.set(sku, {
      product_id: sku,
      name: null,
      sku_id: sku,
      advertised_asin: row.advertised_asin ?? null,
      category: null,
      brand: null,
      inventory_units: null,
      salable: null,
      status: "ADVERTISED",
    });
  }
  return [...bySku.values()];
}

/**
 * One record per Campaign, with the ad product it runs and its measured reach.
 *
 * `baseline_daily_budget` comes from the strategy request when that artifact
 * names one, and is otherwise null: the platform report records spend, not the
 * budget cap that governed it, and presenting observed spend as a budget would
 * misstate what the number is.
 */
function deriveCampaigns(bridgeRows, touchpoints, strategyRequest) {
  const declared = new Map(
    (strategyRequest.campaigns ?? []).map((item) => [item.campaign_id, item]),
  );
  const adProductOf = new Map(
    touchpoints.map((item) => [item.identifier, item.ad_product]),
  );

  const byCampaign = new Map();
  for (const row of bridgeRows) {
    const id = row.campaign_id;
    if (!id) continue;
    if (!byCampaign.has(id)) byCampaign.set(id, []);
    byCampaign.get(id).push(row);
  }

  return [...byCampaign.entries()].map(([id, rows]) => {
    const fromRequest = declared.get(id) ?? {};
    // The Campaign's ad product is whichever its touchpoints report; the
    // reports keep one ad product per Campaign, and the first is taken rather
    // than a guess if that ever stops holding.
    const adProduct =
      fromRequest.ad_product ??
      distinct(
        rows.map((row) => ({
          ad_product: adProductOf.get(
            String(row.touchpoint ?? "").split(":").slice(0, 4).join(":"),
          ),
        })),
        "ad_product",
      )[0] ??
      null;
    return {
      campaign_id: id,
      campaign_name: fromRequest.campaign_name ?? null,
      campaign_group_id: rows[0]?.campaign_group_id ?? null,
      provider: providerOf(adProduct),
      ad_product: adProduct,
      marketplace: rows[0]?.marketplace ?? null,
      baseline_daily_budget: fromRequest.baseline_daily_budget ?? null,
      observed_cost: round(total(rows, "cost"), 2),
      status: fromRequest.status ?? null,
    };
  });
}

/**
 * One record per Ad Group, with the targeting it carries.
 *
 * `initial_daily_budget` is left null: an Ad Group's budget is what the
 * recommendation proposes, not something the historical report states, and the
 * Budget Manager shows the proposal in its own section.
 */
function deriveAdGroups(bridgeRows) {
  const byAdGroup = new Map();
  for (const row of bridgeRows) {
    const id = row.ad_group_id;
    if (!id) continue;
    if (!byAdGroup.has(id)) byAdGroup.set(id, []);
    byAdGroup.get(id).push(row);
  }

  return [...byAdGroup.entries()].map(([id, rows]) => ({
    ad_group_id: id,
    name: null,
    campaign_id: rows[0]?.campaign_id ?? null,
    keyword_count: distinct(rows, "keyword_id").length,
    target_count: distinct(rows, "target_id").length,
    audience_count: distinct(rows, "audience_id").length,
    sku_count: distinct(rows, "sku_id").length,
    allocation_basis: null,
    budget_seed_share: null,
    initial_daily_budget: null,
    status: null,
  }));
}

/**
 * Per-product unit economics, as far as the reports actually establish them.
 *
 * Attributed revenue divided by attributed purchases gives an observed average
 * selling price, which is a real measurement of the account. Cost of goods and
 * the variable costs beneath it are not in any advertising report -- they come
 * from the merchant's own books -- so they stay null, and the contribution
 * margin that depends on them stays null with them rather than being computed
 * from a cost silently assumed to be zero. `margin_source` is therefore left
 * unset: nothing here sourced a margin.
 */
function deriveProductEconomics(bridgeRows, currency) {
  const bySku = new Map();
  for (const row of bridgeRows) {
    const sku = row.sku_id;
    if (!sku) continue;
    const bucket = bySku.get(sku) ?? { revenue: 0, purchases: 0 };
    bucket.revenue += Number(row.assisted_revenue) || 0;
    bucket.purchases += Number(row.assisted_purchase_count) || 0;
    bySku.set(sku, bucket);
  }

  return [...bySku.entries()].map(([sku, totals]) => ({
    product_id: sku,
    currency,
    unit_price:
      totals.purchases > 0 ? round(totals.revenue / totals.purchases, 2) : null,
    unit_cogs: null,
    variable_cost_per_unit: null,
    variable_fulfillment_cost_per_unit: null,
    variable_platform_fee_per_unit: null,
    other_variable_cost_per_unit: null,
    unit_contribution_margin: null,
    margin_source: null,
    observed_revenue: round(totals.revenue, 2),
    observed_purchases: totals.purchases,
  }));
}

/** Which Products each Campaign advertised, as observed in the reports. */
function deriveCampaignProductLinks(bridgeRows) {
  const pairs = new Set();
  const links = [];
  for (const row of bridgeRows) {
    if (!row.campaign_id || !row.sku_id) continue;
    const key = `${row.campaign_id}|${row.sku_id}`;
    if (pairs.has(key)) continue;
    pairs.add(key);
    links.push({ campaign_id: row.campaign_id, product_id: row.sku_id });
  }
  return links;
}

File: backend/repository/master_data.py
"""Derive the master-data catalogue from the committed platform reports.

The Budget Manager's entity sections -- Ad Providers, Products, Campaigns, Ad
Groups, Touchpoints, Product Economics -- describe the account the reports
were pulled from. Those reports are tracked, so the catalogue they imply is
available in every deployment and does not depend on an optional sidecar being
configured.

Derivation rather than a second tracked file is deliberate: a catalogue
committed beside the reports would be free to disagree with them; one read out
of the reports cannot. Where a report does not carry a field -- a Product's
category, a unit price, a Campaign's baseline budget -- the field stays None
rather than being invented, and the interface renders it as missing.

Data flow:
    modules/&#42;/data/simulated/&#42;.csv -> here -> backend/repository/research.py
"""

from __future__ import annotations

from typing import Any, Mapping, Sequence

#: The four Amazon ad products are the vocabulary `mta_common` defines for
#: `Provider.AMAZON_ADS`; anything else is reported under `GENERIC` rather than
#: being dropped, so an added provider surfaces instead of vanishing.
AMAZON_AD_PRODUCTS = frozenset(
    {"SPONSORED_PRODUCTS", "SPONSORED_BRANDS", "SPONSORED_DISPLAY", "AMAZON_DSP"}
)


def _total(rows: Sequence[Mapping[str, Any]], field: str) -> float:
    """Sum a numeric column across rows, treating blanks as zero."""
    running = 0.0
    for row in rows:
        try:
            running += float(row.get(field) or 0)
        except (TypeError, ValueError):
            continue
    return running


def _distinct(rows: Sequence[Mapping[str, Any]], field: str) -> list:
    """Distinct non-empty values of `field`, in first-seen order."""
    seen: list = []
    for row in rows:
        value = row.get(field)
        if value not in (None, "") and value not in seen:
            seen.append(value)
    return seen


def _round(value: float, digits: int) -> float:
    """Round without carrying float noise into the interface."""
    return round(value, digits)


def _split_key(key: str) -> dict:
    """The five-segment touchpoint key's parts.

    `UNSPECIFIED` is the pipeline's marker for a segment the platform does not
    report, and is mapped back to None so the interface shows it as absent
    rather than as a literal value the account actually uses.
    """
    segments = str(key).split(":")
    segments += [None] * (5 - len(segments))
    ad_product, fmt, placement, creative, interaction = segments[:5]

    def clean(value):
        return value if value and value != "UNSPECIFIED" else None

    return {
        "ad_product": ad_product or None,
        "format": clean(fmt),
        "placement": clean(placement),
        "creative": clean(creative),
        "interaction_type": interaction or None,
    }


def _provider_of(ad_product: Any) -> str:
    """The provider a five-segment key belongs to."""
    return "AMAZON_ADS" if ad_product in AMAZON_AD_PRODUCTS else "GENERIC"


def _availability(rows: Sequence[Mapping[str, Any]], field: str) -> str:
    """`AVAILABLE` when any row reports the segment, `NOT_PROVIDED` when none do."""
    return "AVAILABLE" if any(row.get(field) for row in rows) else "NOT_PROVIDED"


def derive_master_data(
    ads_rows: Sequence[Mapping[str, Any]],
    bridge_rows: Sequence[Mapping[str, Any]],
    request: Mapping[str, Any] | None = None,
) -> dict:
    """Build the catalogue.

    Returns the same shape the research sidecar and the database produce, so
    the views cannot tell which source filled it.
    """
    request = request or {}
    touchpoints = _derive_touchpoints(ads_rows)
    campaigns = _derive_campaigns(bridge_rows, touchpoints, request)
    # The account reports one currency; it is read from the data rather than
    # assumed, so a non-USD account is not silently relabelled.
    currencies = _distinct(ads_rows, "currency")
    currency = (
        currencies[0]
        if currencies
        else (request.get("campaign_group") or {}).get("currency")
    )
    return {
        "providers": _derive_providers(touchpoints),
        "products": _derive_products(bridge_rows),
        "campaigns": campaigns,
        "adGroups": _derive_ad_groups(bridge_rows),
        "touchpoints": touchpoints,
        "productEconomics": _derive_product_economics(bridge_rows, currency),
        "campaignProductLinks": _derive_campaign_product_links(bridge_rows),
    }


def _derive_touchpoints(ads_rows: Sequence[Mapping[str, Any]]) -> list[dict]:
    """One record per distinct five-segment key, with its observed economics.

    Cost per click and cost per thousand impressions are computed from the
    report's own totals rather than assumed, and each is reported only for the
    billing type that produced it: a cost-per-mille touchpoint has no
    meaningful cost per click even when a stray click was recorded against it.
    """
    by_key: dict[str, list] = {}
    for row in ads_rows:
        key = row.get("touchpoint") or row.get("normalizedTouchpoint")
        if not key:
            continue
        by_key.setdefault(key, []).append(row)

    by_identifier: dict[str, dict] = {}
    for key, rows in by_key.items():
        parts = _split_key(key)
        # The identifier drops the interaction segment: IMPRESSION and CLICK on
        # one placement are two interactions of a single touchpoint, which is
        # the grouping `supported_interactions` exists to express.
        identifier = ":".join(
            [
                str(parts["ad_product"]),
                parts["format"] or "UNSPECIFIED",
                parts["placement"] or "UNSPECIFIED",
                parts["creative"] or "UNSPECIFIED",
            ]
        )
        record = by_identifier.get(identifier) or {
            "identifier": identifier,
            "provider": _provider_of(parts["ad_product"]),
            "ad_product": parts["ad_product"],
            "format": parts["format"],
            "placement": parts["placement"],
            "placement_availability": (
                "AVAILABLE" if parts["placement"] else "NOT_PROVIDED"
            ),
            "creative": parts["creative"],
            "creative_availability": (
                "AVAILABLE" if parts["creative"] else "NOT_PROVIDED"
            ),
            "interaction_type_availability": "AVAILABLE",
            "supported_interactions": [],
            "impression_enabled": False,
            "click_enabled": False,
            "billing_type": None,
            "cost_per_click": None,
            "cost_per_thousand_impressions": None,
            "base_impressions": None,
            "click_through_rate": None,
            "platform_conversion_rate": None,
            "conversion_log_odds_effect": None,
            "compatibility_keys": [],
            "active": True,
            "_billed": {},
            "_totals": {"impressions": 0.0, "clicks": 0.0, "cost": 0.0},
        }

        interaction = parts["interaction_type"]
        if interaction and interaction not in record["supported_interactions"]:
            record["supported_interactions"].append(interaction)
        record["impression_enabled"] = "IMPRESSION" in record["supported_interactions"]
        record["click_enabled"] = "CLICK" in record["supported_interactions"]
        record["compatibility_keys"].append(key)

        # One touchpoint is commonly billed both ways -- per click on its
        # clicks and per mille on its impressions -- so cost is accumulated
        # against the billing type that produced it rather than against the
        # touchpoint as a whole. Dividing a mixed total by either denominator
        # would report a rate the platform never charged.
        for row in rows:
            cost_type = row.get("cost_type")
            if not cost_type:
                continue
            bucket = record["_billed"].setdefault(
                cost_type, {"cost": 0.0, "clicks": 0.0, "impressions": 0.0}
            )
            bucket["cost"] += float(row.get("cost") or 0)
            bucket["clicks"] += float(row.get("clicks") or 0)
            bucket["impressions"] += float(row.get("impressions") or 0)

        record["_totals"]["impressions"] += _total(rows, "impressions")
        record["_totals"]["clicks"] += _total(rows, "clicks")
        record["_totals"]["cost"] += _total(rows, "cost")
        by_identifier[identifier] = record

    finished = []
    for record in by_identifier.values():
        billed = record.pop("_billed")
        totals = record.pop("_totals")
        per_click = billed.get("CPC")
        per_mille = billed.get("CPM")
        record.update(
            {
                # Both types when the platform bills both, so the row states
                # what it is rather than naming one and hiding the other.
                "billing_type": " + ".join(sorted(billed)) if billed else None,
                # A rate only where cost was actually charged against that
                # denominator. An impression row that delivered volume at no
                # recorded cost yields no rate rather than a rate of zero,
                # which would read as "free".
                "cost_per_click": (
                    _round(per_click["cost"] / per_click["clicks"], 4)
                    if per_click and per_click["cost"] > 0 and per_click["clicks"] > 0
                    else None
                ),
                "cost_per_thousand_impressions": (
                    _round((per_mille["cost"] / per_mille["impressions"]) * 1000, 4)
                    if per_mille
                    and per_mille["cost"] > 0
                    and per_mille["impressions"] > 0
                    else None
                ),
                "observed_impressions": int(totals["impressions"]),
                "observed_clicks": int(totals["clicks"]),
                "observed_cost": _round(totals["cost"], 2),
                # Deliberately not derived. The report records impressions and
                # clicks in separate per-interaction rows that share no
                # denominator, so dividing one by the other does not produce
                # this touchpoint's click-through rate; the observed counts
                # above are what the platform actually states.
                "click_through_rate": None,
                # JavaScript's nullish fallback preserves a measured zero.
                # Do the same here: zero means the report observed no
                # impressions, whereas None means the source had no field.
                "base_impressions": int(totals["impressions"]),
            }
        )
        finished.append(record)
    return finished


def _derive_providers(touchpoints: Sequence[Mapping[str, Any]]) -> list[dict]:
    """One record per provider actually present in the reports."""
    by_provider: dict[str, list] = {}
    for touchpoint in touchpoints:
        by_provider.setdefault(touchpoint["provider"], []).append(touchpoint)
    return [
        {
            "provider": provider,
            "supported_ad_products": _distinct(rows, "ad_product"),
            "format_availability": _availability(rows, "format"),
            "placement_availability": _availability(rows, "placement"),
            "creative_availability": _availability(rows, "creative"),
            "interaction_type_availability": "AVAILABLE",
            "active": True,
        }
        for provider, rows in by_provider.items()
    ]


def _derive_products(bridge_rows: Sequence[Mapping[str, Any]]) -> list[dict]:
    """One record per advertised stock-keeping unit.

    The reports identify a product by its Amazon Standard Identification Number
    and its stock-keeping unit and carry nothing else about it. Those fields
    are present and None, which is what lets the Form editor show the whole
    shape of a Product while being honest that the report does not describe it.
    """
    by_sku: dict[str, dict] = {}
    for row in bridge_rows:
        sku = row.get("sku_id")
        if not sku or sku in by_sku:
            continue
        by_sku[sku] = {
            "product_id": sku,
            "name": None,
            "sku_id": sku,
            "advertised_asin": row.get("advertised_asin"),
            "category": None,
            "brand": None,
            "inventory_units": None,
            "salable": None,
            "status": "ADVERTISED",
        }
    return list(by_sku.values())


def _derive_campaigns(
    bridge_rows: Sequence[Mapping[str, Any]],
    touchpoints: Sequence[Mapping[str, Any]],
    request: Mapping[str, Any],
) -> list[dict]:
    """One record per Campaign, with the ad product it runs and its reach.

    `baseline_daily_budget` comes from the strategy request when that artifact
    names one, and is otherwise None: the platform report records spend, not
    the budget cap that governed it, and presenting observed spend as a budget
    would misstate what the number is.
    """
    declared = {
        item.get("campaign_id"): item for item in (request.get("campaigns") or [])
    }
    ad_product_of = {item["identifier"]: item["ad_product"] for item in touchpoints}

    by_campaign: dict[str, list] = {}
    for row in bridge_rows:
        campaign_id = row.get("campaign_id")
        if not campaign_id:
            continue
        by_campaign.setdefault(campaign_id, []).append(row)

    campaigns = []
    for campaign_id, rows in by_campaign.items():
        from_request = declared.get(campaign_id) or {}
        # The Campaign's ad product is whichever its touchpoints report; the
        # reports keep one ad product per Campaign, and the first is taken
        # rather than a guess if that ever stops holding.
        observed = _distinct(
            [
                {
                    "ad_product": ad_product_of.get(
                        ":".join(str(row.get("touchpoint") or "").split(":")[:4])
                    )
                }
                for row in rows
            ],
            "ad_product",
        )
        ad_product = from_request.get("ad_product") or (observed[0] if observed else None)
        campaigns.append(
            {
                "campaign_id": campaign_id,
                "campaign_name": from_request.get("campaign_name"),
                "campaign_group_id": rows[0].get("campaign_group_id"),
                "provider": _provider_of(ad_product),
                "ad_product": ad_product,
                "marketplace": rows[0].get("marketplace"),
                "baseline_daily_budget": from_request.get("baseline_daily_budget"),
                "observed_cost": _round(_total(rows, "cost"), 2),
                "status": from_request.get("status"),
            }
        )
    return campaigns


def _derive_ad_groups(bridge_rows: Sequence[Mapping[str, Any]]) -> list[dict]:
    """One record per Ad Group, with the targeting it carries.

    `initial_daily_budget` is left None: an Ad Group's budget is what the
    recommendation proposes, not something the historical report states, and
    the Budget Manager shows the proposal in its own section.
    """
    by_ad_group: dict[str, list] = {}
    for row in bridge_rows:
        ad_group_id = row.get("ad_group_id")
        if not ad_group_id:
            continue
        by_ad_group.setdefault(ad_group_id, []).append(row)

    return [
        {
            "ad_group_id": ad_group_id,
            "name": None,
            "campaign_id": rows[0].get("campaign_id"),
            "keyword_count": len(_distinct(rows, "keyword_id")),
            "target_count": len(_distinct(rows, "target_id")),
            "audience_count": len(_distinct(rows, "audience_id")),
            "sku_count": len(_distinct(rows, "sku_id")),
            "allocation_basis": None,
            "budget_seed_share": None,
            "initial_daily_budget": None,
            "status": None,
        }
        for ad_group_id, rows in by_ad_group.items()
    ]


def _derive_product_economics(
    bridge_rows: Sequence[Mapping[str, Any]], currency: Any
) -> list[dict]:
    """Per-product unit economics, as far as the reports actually establish them.

    Attributed revenue divided by attributed purchases gives an observed
    average selling price, which is a real measurement of the account. Cost of
    goods and the variable costs beneath it are not in any advertising report
    -- they come from the merchant's own books -- so they stay None, and the
    contribution margin that depends on them stays None with them rather than
    being computed from a cost silently assumed to be zero. `margin_source` is
    therefore left unset: nothing here sourced a margin.
    """
    by_sku: dict[str, dict] = {}
    for row in bridge_rows:
        sku = row.get("sku_id")
        if not sku:
            continue
        bucket = by_sku.setdefault(sku, {"revenue": 0.0, "purchases": 0.0})
        bucket["revenue"] += float(row.get("assisted_revenue") or 0)
        bucket["purchases"] += float(row.get("assisted_purchase_count") or 0)

    return [
        {
            "product_id": sku,
            "currency": currency,
            "unit_price": (
                _round(totals["revenue"] / totals["purchases"], 2)
                if totals["purchases"] > 0
                else None
            ),
            "unit_cogs": None,
            "variable_cost_per_unit": None,
            "variable_fulfillment_cost_per_unit": None,
            "variable_platform_fee_per_unit": None,
            "other_variable_cost_per_unit": None,
            "unit_contribution_margin": None,
            "margin_source": None,
            "observed_revenue": _round(totals["revenue"], 2),
            "observed_purchases": totals["purchases"],
        }
        for sku, totals in by_sku.items()
    ]


def _derive_campaign_product_links(
    bridge_rows: Sequence[Mapping[str, Any]],
) -> list[dict]:
    """Which Products each Campaign advertised, as observed in the reports."""
    pairs: set[tuple] = set()
    links = []
    for row in bridge_rows:
        campaign_id = row.get("campaign_id")
        sku = row.get("sku_id")
        if not campaign_id or not sku:
            continue
        if (campaign_id, sku) in pairs:
            continue
        pairs.add((campaign_id, sku))
        links.append({"campaign_id": campaign_id, "product_id": sku})
    return links
